### PR TITLE
Rewrite ReduceL1, ReduceL2, ReduceLogSum, ReduceLogSumExp, ReduceSumSquare in the ONNX dialect

### PR DESCRIPTION
**Pull request by tungld**
Given that there are ops in the ONNX dialect that can be rewritten by using other "primitive" ONNX ops, this patch is to rewrite some of them, i.e. ReduceL1, ReduceL2, ReduceLogSum, ReduceLogSumExp, ReduceSumSquare. Rewriting will be implemented by using canonicalization pass.

This patch is also changing the order of calling shape-inference and canonicalization passes, making the canonicalization pass be called before the shape-inference pass. The reason is to avoid implementing shape inference for ops that are rewritten.

### CODE REVIEW

**Comment by tjingrant** [comment]
@tungld wonder if you can give https://github.com/clang-ykt/ONNF/pull/29 a review. It's a small patch that's blocking all the lowering patches. I hope to merge it ASAP.

**Comment by tjingrant** [comment]
The buildbot failure you see here will also go away once #29 is merged.

**Comment by tungld** [comment]
@tjingrant OK, I will review it this afternoon.

File: src/pass/onnx_rewrite.cpp [hunk 1]
```diff
@@ -0,0 +1,314 @@
...
+// More information about the limitation can be found here:
+// https://github.com/llvm/llvm-project/blob/master/mlir/docs/DeclarativeRewrites.md#building-operations
+//
+
+//===----------------------------------------------------------------------===//
+// ONNXReduceL1Op %X = ONNXReduceSumOp (ONNXAbsOp %X)
+//===----------------------------------------------------------------------===//
+struct ReduceL1OpPattern : public RewritePattern {
+  ReduceL1OpPattern(MLIRContext *context)
+      : RewritePattern("onnx.ReduceL1", {"onnx.Abs", "onnx.ReduceSum"}, 1,
```
Comment (tjingrant) [inline src/pass/onnx_rewrite.cpp]: Any chance we can get the operation names by invoking some method of these operations? Like `ONNXReduceL1Op::name()`? Using hard-coded operation names like such is a bit risky.

File: src/pass/onnx_rewrite.cpp [hunk 2]
```diff
@@ -0,0 +1,314 @@
...
+
+using namespace mlir;
+
+namespace {
+
+// Due to a limitation of ODS that "all ODS-generated `build()` methods require
+// specifying the result type(s), unless the op has known traits like
+// `SameOperandsAndResultType` that we can use to auto-generate a `build()`
+// method with result type deduction", we will to rewrite some patterns manually
+// until we find a better way.
```
Comment (tjingrant) [inline src/pass/onnx_rewrite.cpp]: I'm a bit confused, why the quoted text implies that we have to rewrite patterns manually? Can't we generate build() method ourselves? c.f. https://github.com/clang-ykt/ONNF/blob/6959cf45869ddce0e960820788ec3358b4f16b21/src/dialect/krnl/krnl_ops.td#L30 .

Comment (tungld) [reply src/pass/onnx_rewrite.cpp]: Sorry for the confusion. I think it does not mean we have to. We can use custom builders like what you said. Here, I write them manually due to two reasons:
- To insert custom builders in TableGen, it is better to change the script `gen_doc.py` to generate all possibles custom builders for a large class of operations. At the time of this patch created, the `gen_doc.py` was changing, so I decided to write manually to reduce conflicts.
- The use of custom builders is to help rewrite in a declarative way. In declarative rewriting, we should deal with optional attributes. E.g. for reduction ops, attribute `axes` is optional. I guess we should declare multiple patterns to handle optional attributes, but I haven't tried it yet.

File: src/pass/onnx_rewrite.cpp (hunk 1 above)
Comment (tungld) [reply src/pass/onnx_rewrite.cpp]: I revised the code. Thanks!

File: src/pass/onnx_rewrite.cpp [hunk 3]
```diff
@@ -0,0 +1,314 @@
...
+      : RewritePattern("onnx.ReduceL1", {"onnx.Abs", "onnx.ReduceSum"}, 1,
+                       context) {}
+
+  PatternMatchResult matchAndRewrite(Operation *op,
+                                     PatternRewriter &rewriter) const override {
+    // Variables for capturing values and attributes used for creating ops
+    Operation::operand_range X(op->getOperands());
+
+    // Match
+    auto opODS = dyn_cast_or_null<ONNXReduceL1Op>(op);
```
[the remark this answers is not in the document]
Comment (tungld) [reply src/pass/onnx_rewrite.cpp]: I revised the code. Thanks!

**Comment by tjingrant** [comment]
@tungld the buildbot error is likely caused by an out-of-memory issue. The MLIR cache seems to have been automatically invalidated on the buildbot, I'm looking into this problem.

File: src/pass/onnx_rewrite.cpp (hunk 2 above)
Comment (tungld) [reply src/pass/onnx_rewrite.cpp]: Done. I have updated the comments.

**Pull request merged**